nifi-nar-bundles/nifi-azure-bundle/nifi-azure-reporting-task/src/main/java/org/apache/nifi/reporting/azure/loganalytics/AzureLogAnalyticsReportingTask.java

```diff
@@ -0,0 +1,335 @@
+                    allMetrics.addAll(collectMetrics(instanceId, status, processGroupName, jvmMetricsCollected));
+                }
+            }
+            HttpPost httpPost = getHttpPost(urlEndpointFormat, workspaceId, logName);
+            sendMetrics(httpPost, workspaceId, linuxPrimaryKey, allMetrics);
+        } catch (IOException | RuntimeException e) {
+            getLogger().error("Exception in outmost block", e);
+        }
+    }
+
+    /**
+     *  Construct HttpPost and return it
+     * @param urlFormat URL format to Azure Log Analytics Endpoint
+     * @param workspaceId your azure log analytics workspace id
+     * @param logName log table name where metrics will be pushed
+     * @return HttpsURLConnection to your azure log analytics workspace
+     * @throws IllegalArgumentException if dataCollectorEndpoint url is invalid
+     */
+    protected HttpPost getHttpPost(final String urlFormat, final String workspaceId, final String logName)
+        throws IllegalArgumentException {
+        String dataCollectorEndpoint =
+            MessageFormat.format(urlFormat, workspaceId);
+        HttpPost post = new HttpPost(dataCollectorEndpoint);
+        post.addHeader("Content-Type", "application/json");
+        post.addHeader("Log-Type", logName);
+        return post;
+    }
+    /**
+     * send collected metrics to azure log analytics workspace
+     * @param request HttpPost to Azure Log Analytics Endpoint
+     * @param workspaceId your azure log analytics workspace id
+     * @param linuxPrimaryKey your azure log analytics workspace key
+     * @param allMetrics collected metrics to be sent
+     * @throws IOException when there is an error in https url connection or read/write to the onnection
+     * @throws IllegalArgumentException when there a exception in converting metrics to json string with Gson.toJson
+     * @throws RuntimeException when httpPost fails with none 200 status code
+     */
+    protected void sendMetrics(final HttpPost request, final String workspaceId, final String linuxPrimaryKey, final List<Metric> allMetrics)
+            throws IOException, IllegalArgumentException, RuntimeException {
+
+        Gson gson = new GsonBuilder().create();
+        StringBuilder builder = new StringBuilder();
+        builder.append('[');
+        for (Metric current : allMetrics) {
+            builder.append(gson.toJson(current));
+            builder.append(',');
+        }
+        builder.append(']');
+
+        final String rawJson = builder.toString();
+        final int bodyLength = rawJson.getBytes(UTF8).length;
+        final String nowRfc1123 = RFC_1123_DATE_TIME.format(ZonedDateTime.now(ZoneOffset.UTC));
+        final String createAuthorization = createAuthorization(workspaceId, linuxPrimaryKey, bodyLength, nowRfc1123);
+        request.addHeader("Authorization", createAuthorization);
+        request.addHeader("x-ms-date", nowRfc1123);
+        request.setEntity(new StringEntity(builder.toString()));
```

Comment:
We already have the result of calling `builder.toString()` stored in `rawJson` - we should use that rather than creating another copy of the String.

Reply:
fixed as recommended